Try: Add simplified readme validator

carolinan opened the pull request:
Fixes https://github.com/WordPress/theme-check/issues/375

This check and the accompanying parser is copied from the plugin readme validator.

It is more limited than the plugin readme validator, and it does not work for themes that are in sub folders.
The subfolder issue may be solved by #370. Untested.

**How to test:**

1. Test with a theme that is missing a readme.txt and readme.md
2. Confirm that the error about the missing readme is displaying.
3. Test with a theme with missing headers in the readme.
4. Confirm that the warnings are displaying.
5. Test with a theme with incorrect headers in the readme.
6. Confirm that the warnings are displaying.

carolinan committed:
Try: Add simplified readme validator

carolinan commented:
The report:
![Theme check displays a result with missing headers in the readme file](https://user-images.githubusercontent.com/7422055/127489193-eb25306c-9f5f-4475-ba1e-f89777cf9ef1.png)

carolinan committed:
Fix CS issues

carolinan committed:
Replace short array syntax

Replace short array syntax
Remove the readme.txt file from class-file-check.php
Update the error message when the file is missing.

carolinan committed:
Fix coding standard issue.

carolinan committed:
Update class-readme-check.php

carolinan committed:
more CS issues..

TeBenachi commented:
Step 2. Tested and can confirm the missing error.  

Step 4. Tested and confirmed.  

Step 6. Tested and confirmed. Screenshot attached. 

One thing I noticed. There was no warning when I added multiple contributors. 

![screenshot2](https://user-images.githubusercontent.com/30734209/128953640-be7611d7-406d-4fe2-8ae7-b3d17bc82aa6.jpg)

carolinan commented:
Thank you.
Yes good point about the contributors. I left that out since the plugin can not verify the user name (which the plugin readme validator can).

kafleg commented:
**Test with a theme that is missing a readme.txt and readme.md** 
Tested and worked. Also, i liked this _REQUIRED: Found a copy of Underscores. See README.md. Learn how to update the files for your own theme._ If there is README.md file of default underscores found. 

**Confirm that the error about the missing readme is displaying.**
Confirmed - _REQUIRED: The readme file is missing._

**Test with a theme with missing headers in the readme.**
Confirmed. 
```
README WARNING: The Tested up to field is missing from the readme.
README WARNING: The Requires at least field is missing from the readme.
README WARNING: The Requires PHP field is missing from the readme.
README WARNING: The License field is missing from the readme.
README WARNING: The License URI field is missing from the readme.
README WARNING: The Contributors field is missing from the readme.
```

**Confirm that the warnings are displaying.**
Done, please check above. 

**Test with a theme with incorrect headers in the readme.**
Only one issue shown, `README WARNING: The Requires at least field in the readme was ignored. This field should only contain a valid WordPress version such as 0.0 or -0.1.`

my readme header is like below,
```
Contributors: templatesellxyz
Tags: red, cat, dog
Requires at least: 22.5
Requires PHP: 9.0
Tested up to: 4.3
Stable tag: 1.2.9
License: License.txt
License URI: wordpress.org
```

**Confirm that the warnings are displaying.**
Check above. 


On a side note, is this INFO necessary? 
```
README INFO: No == Frequently Asked Questions == section was found in the readme.
README INFO: No == Description == section was found in the readme.
README INFO: No == Changelog == section was found in the readme.
```

carolinan commented:
The requires at least is compared to ` $latest_wordpress_version = '5.8';` so 22.5 should fail. How can we document this?

carolinan commented:
The info does make it noisy. My thought was to:
-Match the plugin validator info
-Prepare for when the info is shown in the directory.

But if you prefer it can be removed and reconsidered once there are solid plans to show it in the directory.

TeBenachi commented:
How about something like this? 

The Requires at least field requires the least WordPress version. This field should only contain a valid WordPress version such as 0.0 or 0.1. 

I didn’t notice it yesterday but there is “ - " (minus sign) in the Requires at least warning message?  

![screenshot3](https://user-images.githubusercontent.com/30734209/129020430-fde04dbd-9ac5-478e-8c02-db7cfdb3542c.jpg)

carolinan commented:
Yeah I think it is broken. I'd like to compare with the live plugin validator.

carolinan committed:
Update class-readme-check.php

kafleg commented:
Looks good. 

`README WARNING: The Contributors field should only contain one WordPress.org username. Remember that usernames are case-sensitive.`

TeBenachi commented:
`Contributors: Ginger Snap`

When I added 2 users without “,” comma between them, the check returned no error. 
I know most of the people use “,” comma to separate stings but should we also add a check in case? 

`Contributors: Ginger, Snap` worked perfectly.

carolinan commented:
I have not used the sniffer for so long, that I forgot that it checks for this:
https://github.com/WPTT/theme-sniffer/blob/development/src/sniffs/readme/class-contributors.php

carolinan commented:
> protected $latest_wordpress_version = '5.8';
> 
> We need to make it dynamic.

But how?

TeBenachi commented:
> I have not used the sniffer for so long, that I forgot that it checks for this:
> https://github.com/WPTT/theme-sniffer/blob/development/src/sniffs/readme/class-contributors.php

Thanks for the link. I got the contributor validation working based on the sniffer. It’s silly but I don’t know how to create a branch off a branch.  Any suggestions on how to proceed.

dd32 commented on checks/class-readme-check.php:
```diff
@@ -0,0 +1,241 @@
+		 *
+		 * @var string $latest_wordpress_version
+		 */
+		$latest_wordpress_version = '5.8';
```
```suggestion
		if ( defined( 'WP_CORE_LATEST_RELEASE' ) ) {
			// When running on WordPress.org, this constant defines the latest WordPress release.
			$latest_wordpress_version = WP_CORE_LATEST_RELEASE;
		} else {
			// Assume that the local environment being tested in is up to date.
			$latest_wordpress_version = $GLOBALS['wp_version'];
		}
```

carolinan committed:
Update checks/class-readme-check.php

Co-authored-by: Dion Hulse <dion@wordpress.org>

kafleg commented:
@dd32 What will happen if my local WordPress version is less than the current version?

kafleg commented:
No notice is shown in this case, `Tested up to: 5.9` 

But if we go above `6.0` then it shows,` README WARNING: The Tested up to field in the readme was ignored. This field should only contain a valid WordPress version such as 5.8 or 5.9.`

If the WordPress version will be higher than 5.9, then do we need to reset the version again?

Nothing shows for, `Requires PHP: 22.5`, here 22.5 is just a testing value. 

Rest of the things looks fine to me.

kafleg commented:
I found one thing,
If there is space in readme header items, then it shows the errors. 

![image](https://user-images.githubusercontent.com/13062231/129438333-fec227a7-f538-4b0b-88f5-a5f4705c7c20.png)

And errors list, 
![image](https://user-images.githubusercontent.com/13062231/129438346-66e8d0e0-3e05-414a-ac85-90f7e059f9c2.png)

Once I removed that space, only the error of contributors is displayed. 
![image](https://user-images.githubusercontent.com/13062231/129438365-8ae982fe-ccc7-48b9-bbf3-46ff6815d41f.png)

carolinan commented:
> If the WordPress version will be higher than 5.9, then do we need to reset the version again?
We would have had to but Dion's update solves that.

kafleg commented:
> > If the WordPress version will be higher than 5.9, then do we need to reset the version again?
> 
> We would have had to but Dion's update solves that.

Okay, thanks for the update.

Any idea how to fix the readme issues with spacing?

carolinan commented:
> > I have not used the sniffer for so long, that I forgot that it checks for this:
> > https://github.com/WPTT/theme-sniffer/blob/development/src/sniffs/readme/class-contributors.php
> 
> Thanks for the link. I got the contributor validation working based on the sniffer. It’s silly but I don’t know how to create a branch off a branch. Any suggestions on how to proceed.

I am not sure either 😊 Maybe we need to do it in two steps and merge this one first?

carolinan commented:
> Any idea how to fix the readme issues with spacing?
I'll see if I can make it remove blank lines from the header, but the warning is also correct, the blank line must not be there.

carolinan committed:
Check if the name in the readme matches the theme name

Check if the name in the readme matches the theme name, if not, error.
Update the $latest_wordpres_version variable.
Remove empty lines to prevent errors with the check.

carolinan committed:
Solve CS problems

carolinan commented:
https://github.com/WordPress/theme-check/pull/376/commits/8e696d2479a9e4ff7ddf01467bd5a9ba6847973c Checks if the name in the readme matches the theme name, if not, error.
Updates the $latest_wordpres_version variable.
Removes empty lines to prevent errors with the check.

To avoid reading the wrong readme file if there is more than one readme, this PR needs to be tested together with
 #370.

carolinan committed:
Fix CS problems

kafleg commented:
Theme readme file,
![image](https://user-images.githubusercontent.com/13062231/130052402-b23f2747-4b51-47f7-a3a5-6054cd252861.png)

But there is no error/notice shown.
![image](https://user-images.githubusercontent.com/13062231/130052458-72e24455-7290-4b95-92b7-5cfab80e599e.png)

I checked again with removing other,
![image](https://user-images.githubusercontent.com/13062231/130052787-14c17727-9f31-4c3d-ad9e-e5d466e5df42.png)

Now, I can see an error notice for Tested up to, Requires at least and Requires PHP.
![image](https://user-images.githubusercontent.com/13062231/130052825-317acbe3-8bfd-4ae6-9fd6-e5c6569814a3.png)

carolinan commented:
Does the license and license URI have a single space after the colon?

kafleg commented:
Requires at least: 5.5
Tested up to: 5.8
Requires PHP: 7.4 
Stable tag: 1.0.0 
License:  
License URI:  

Tested with both adding space, removing space, double space. The result is same.

carolinan committed:
Merge pull request #393 from WordPress/master

Sync branch with master

carolinan commented:
I think I found the bug where it does not show that the license is missing, but I need your full readme file to confirm.

It is not only checking the file header, so if the words "License: blah blah" are elsewhere in the readme file, like for an asset, it will use that.

carolinan committed:
Fix for 392

Try to solve timeout issues.
Check for uppercase file names as well as lowercase with stripos.

carolinan committed:
fix coding standard issues
